[Merged by Bors] - Altair consensus changes and refactors

michaelsproul opened the pull request:
## Proposed Changes

Implement the consensus changes necessary for the upcoming Altair hard fork.

## Additional Info

This is quite a heavy refactor, with pivotal types like the `BeaconState` and `BeaconBlock` changing from structs to enums. This ripples through the whole codebase with field accesses changing to methods, e.g. `state.slot` => `state.slot()`.

michaelsproul committed:
Start Altair refactor

michaelsproul commented:
Current status:

- `make` compiles a Lighthouse binary but it doesn't yet work (it should be close)
- Actual Altair consensus code is stubbed out
- Many tests are broken due to the refactor
- Remaining tasks are tagged `FIXME(altair)` in the code

michaelsproul commented:
https://github.com/sigp/lighthouse/issues/2278 should likely use this branch as a base. We could keep this branch clean with PRs + rebases, and then merge it without squashing. Or we could split consensus/networking changes into two PRs, and merge them one after the other.

michaelsproul committed:
Add AltairConfig

michaelsproul commented:
Lighthouse can now run against mainnet/pyrmont from this branch, I've temporarily hardcoded the fork slot to INT_MAX.

Will work on the Altair EF tests + consensus tomorrow.

michaelsproul committed:
Towards SSZ tests passing

SignedBeaconBlock failing due to INT_MAX slot issue
TreeHashing for BeaconState failing
Cached tree hashing for BeaconState currently disabled

michaelsproul committed:
Generate as_base/as_altair with superstruct

michaelsproul committed:
Most SSZ static tests passing

michaelsproul commented:
TODO:

- Change `SignedBeaconBlock` to a `superstruct` so that it's possible to hint which variant we want manually. This is necessary for @pawanjay176's networking changes.

Also a note for later:

- We'll need something analogous to the op pool to collect `SignedContributionAndProof`s from the network and aggregate them for use in proposed blocks. This should likely live on the `BeaconChain`. Spec here: https://github.com/ethereum/eth2.0-specs/blob/v1.1.0-alpha.2/specs/altair/validator.md#preparing-a-beaconblock

realbigsean committed:
justification_and_finalization changes

realbigsean committed:
refactor to match eth2  spec, and fix epoch processing EF tests

realbigsean committed:
Merge branch 'consensus-spec-tests' of https://github.com/realbigsean/lighthouse into altair-process-epoch

 Conflicts:
	consensus/state_processing/src/per_epoch_processing/base.rs

michaelsproul committed:
Make SignedBeaconBlock a superstruct

michaelsproul committed:
Merge remote-tracking branch 'origin/unstable' into altair

michaelsproul committed:
Fix ef_tests for SBB change

michaelsproul commented on consensus/types/src/signed_beacon_block.rs:
```diff
@@ -75,40 +147,41 @@ impl<E: EthSpec> SignedBeaconBlock<E> {
-        self.message.tree_hash_root()
+        match self {
+            SignedBeaconBlock::Base(block) => block.message.tree_hash_root(),
+            SignedBeaconBlock::Altair(block) => block.message.tree_hash_root(),
```
Note to self: this doesn't need the `match` any more

realbigsean committed:
refactors

realbigsean committed:
Merge branch 'altair' of https://github.com/sigp/lighthouse into altair-process-epoch

realbigsean committed:
refactor justification and finalization, begin altair rewards and penalties

realbigsean committed:
get rewards and penalties working for altair

realbigsean committed:
fix some imports

michaelsproul committed:
Implement basic sync committee functions

michaelsproul committed:
Merge remote-tracking branch 'sean/altair-process-epoch' into altair

michaelsproul committed:
Implement process_sync_committee

michaelsproul committed:
Update EF operation tests incl sync agg

realbigsean committed:
small refactors

michaelsproul committed:
Update attestation/deposit processing

michaelsproul committed:
Implement `upgrade_to_altair`, genesis init test

realbigsean committed:
Merge branch 'altair' of https://github.com/sigp/lighthouse into more-altair-epoch-processing

 Conflicts:
	consensus/state_processing/src/per_epoch_processing/altair/rewards_and_penalties.rs

realbigsean committed:
fix slashings and implement sync committee processing

realbigsean committed:
refactor common epoch processing methods

realbigsean committed:
delete unused imports

michaelsproul committed:
Fix typo in sync_committee_updates

michaelsproul committed:
Update to v1.1.0-alpha.3

michaelsproul committed:
Fix remaining EF tests/handlers!

michaelsproul commented:
It's looking good! All the consensus tests from the EF v1.1.0-alpha.3 are passing now.

The upcoming changes I have in mind are:

- [x] Review + merge Pawan's networking changes
- [x] Pools for `SignedContributionAndProof`s and other sync aggregate objects that can come off the network. See [Altair validator spec](https://github.com/ethereum/eth2.0-specs/blob/v1.1.0-alpha.3/specs/altair/validator.md). This will likely be similar to the naive aggregation pool (and maybe the op pool), accepting items from the network via methods on the `BeaconChain`.
- [x] Abstract block proposal to propose Altair blocks (using above pool)
- [x] Adapt attestation packing in op pool to use Altair rewards for calculations (where relevant)
- [x] Annihilate those `TestingXBuilder`s. I reckon we throw them in the bin (and those fragile tests based on them). IMO the way forward is to use the `BeaconChainHarness` to generate attestations/slashings/etc for testing, because it has a lot more context to make that ergonomic + sensible. In other places where the existing builders are used, I think we can probably find alternatives that are less cumbersome (use `TestRandom` or use the `BeaconChainHarness`).
- [x] Get the `beacon_chain` tests passing (see previous)
- [x] Update `superstruct` to generate _partial getters_ for fields shared by some variants. I think this is preferable to the `as_base` and `as_altair` methods we use at the moment, because it will continue to work for future forks retaining the same fields (i.e. the merge). I've started working on this locally and will push to `sigp/superstruct` soon (TM).
- [x] Resolve the `ValidatorStatus` conundrum (see comments in `per_epoch_processing::altair`) [assigned: @michaelsproul]
- [x] Add a service to the validator client to perform sync committee duties. Requires the API from the next item. See [Altair validator spec](https://github.com/ethereum/eth2.0-specs/blob/v1.1.0-alpha.3/specs/altair/validator.md#sync-committees).
- [x] Add APIs to the beacon node for fetching & broadcasting sync committee signatures and aggregates. We should check if a standard API has been proposed, and consider proposing one if not. Forming aggregates will depend on the pool(s) for `SignedContributionAndProof`s (item #2).
- [x] Plug in `upgrade_to_altair`
- [x] Add EF test runner for `upgrade_to_altair` (`fork` subdir)
- [x] Parametrise updated tests (`beacon_chain`, etc) by fork, possibly using env var approach for simplicity
- [x] Update `lcli` commands (`new_testnet` could be useful for running local Altair testnets soon)
- [ ] Update aggregate attestation gossip filter: https://github.com/ethereum/eth2.0-specs/pull/2225
- [x] Make sure to update or replace `BeaconChain::duration_to_next_fork` (spotted by Pawan)

Stretch goals:

- [ ] Implement runners for the EF fork choice tests (haven't looked into this).
- [x] Do some profiling/benchmarking and consider stripping out/caching proposer indices & total balances (links in with validator status considerations)
- [ ] Try supercharging tree hashing by wrapping `validators_mut()` so that it records the dirty indices.
- [ ] Strip all `/// Spec vX.Y.Z` comments (they're just a maintenance burden, and hugely out of date)

I've probably forgotten a few things, if anyone notices any, please post here!

realbigsean committed:
Fix beacon chain tests non-altair (#2305)

michaelsproul committed:
Use RefMut

michaelsproul committed:
Reset Cargo.lock to "stable" style

michaelsproul committed:
Tidy block tree hash

michaelsproul committed:
Merge remote-tracking branch 'origin/unstable' into altair

michaelsproul committed:
Temporarily hardcode base fork in op pool

michaelsproul committed:
Update block proposal & op pool

michaelsproul committed:
Fix bounds check on compute_sync_committee_indices

michaelsproul committed:
Tidy phase0 op pool

michaelsproul committed:
Remove global fork schedule (#2314)

* Remove global fork schedule
* Fix remaining EF tests for refactor
* Get top-level compiling again
* Bubble YAML config changes

michaelsproul committed:
Use superstruct partial getters

This makes our Altair consensus logic compatible with future forks.

michaelsproul committed:
More test cleanups and fixes

realbigsean committed:
Lots of test refactors (#2315)

* Fix beacon chain tests non-altair

* fix op pool tests

* fix merge

* get block processing tests working

* Refactor state transition vector tests, half are passing

* cargo fmt

* PR updates

* Skip ahead to epoch 256 on beacon harness tests. Fix op pool tests.

* clippy updates

michaelsproul committed:
Use `upgrade_to_altair` in state transition

michaelsproul committed:
More test fixes

michaelsproul commented:
I've updated the "roadmap"/TODO list above.

I have an idea for how to resolve the validator status thing, so I'll tentatively take that for tomorrow. The idea is: translate the _old_ validator status into _new_ participation bits, and re-jig the validator monitor to mostly use the new stuff (with `legacy: Option<ValidatorStatus>` in there as well until the fork happens).

michaelsproul committed:
Add fork upgrade test, fix EF spec test

michaelsproul commented:
Draft API spec from JGM that we can provide feedback on in the #apis channel of Eth R&D: https://hackmd.io/@QYHAVYiHRg65pdI_CEm7Eg/syncommitteeapi

michaelsproul committed:
Update lcli and local testnet scripts

michaelsproul committed:
More test updates, release-tests might pass now...

michaelsproul commented:
Need to mark `state_transition` tests as `release` only

We've also added ~10-15 mins to the release tests. I think it may make sense to split it into `beacon_chain` and non-`beacon_chain` tests (also so we can run the beacon_chain tests for multiple fork configs)

michaelsproul committed:
Nuke old store benchmark

michaelsproul committed:
Fix arbitrary instances

michaelsproul committed:
Participation bits fixes, clippy fixes

michaelsproul commented:
I noticed we introduced some new indexing into vecs during Altair. I might add a clippy lint against this and take out all the indexing, depending on how much work that seems to be: https://rust-lang.github.io/rust-clippy/stable/index.html#indexing_slicing

michaelsproul committed:
Disable indexing in types & state processing

michaelsproul committed:
Delete tree hash and state proc benchmarks

michaelsproul committed:
Disable aggressive lints in tests

michaelsproul committed:
Appease udeps

michaelsproul committed:
Fix the RPC tests without hacks

Thanks @pawanjay176

michaelsproul committed:
Fix double update in github cargo fmt

michaelsproul committed:
Use published superstruct

michaelsproul committed:
Try splitting beacon_chain tests out of CI

michaelsproul committed:
Fix cargo-fmt --all

michaelsproul commented:
This is ready for review, modulo me getting CI to pass in a more timely fashion. Unfortunately my idea of splitting the tests was a bit of a non-starter, because most of the time (49 minutes) is spent on compilation!

Will keep investigating

michaelsproul committed:
Abstract over fork in beacon_chain tests

Also consolidate persistence_tests into store tests to
reduce the # of integration test binaries slightly.

michaelsproul committed:
Fix clippy

michaelsproul committed:
Merge remote-tracking branch 'origin/unstable' into altair

michaelsproul committed:
Maintain backwards compatibility with std config

michaelsproul commented:
Commit c89de3ba6d9f232aa51c51f62ce81aef044583f4 is running on the SigP Pyrmont validators. I've also tested that commit with Infura's Pyrmont node, and it is working well (the Teku BN doesn't report any Altair-related fields on Pyrmont :+1: )

michaelsproul commented:
The sync committee cache needs to removed in light of https://github.com/ethereum/eth2.0-specs/issues/2392, but that should be a fairly self-contained change and I think this PR is still otherwise ready for review.

paulhauner commented:
I'm starting review here, *finally* :tada: 

It's a thicc boi so I'm going to do a few passes. On these initial passes I'll just leave some comments (instead of a "review"). Hopefully it's not too noisy!

paulhauner commented on consensus/types/src/beacon_state.rs:
```diff
@@ -191,52 +225,88 @@ where
     #[tree_hash(skip_hashing)]
     #[test_random(default)]
+    #[derivative(Clone(clone_with = "clone_default"))]
+    pub current_sync_committee_cache: SyncCommitteeCache,
```
I noticed that this exists in `Base`, not just `Altair`. Is that intentional?

michaelsproul replied:
It was for simplicity, but this entire cache will be deleted because it turns out to be impossible to implement correctly: ethereum/eth2.0-specs#2392

I'll do that today and push an update

paulhauner committed:
Provide clarity about BeaconState leaf count (#2339)

michaelsproul committed:
Remove sync committee indices cache

We could add back a cache at some point, but it would likely look
completely different due to having to compute the indices "backwards"
from the public keys. Dealing with all the cases involving rebuilding
the cache efficiently, or what to do if it isn't built will be quite
involved so I think it's better we wait to see if we need it before
optimising prematurely.

paulhauner commented on consensus/types/src/beacon_state.rs:
```diff
@@ -191,52 +222,81 @@ where
-    /// Spec v0.12.1
     pub fn new(genesis_time: u64, eth1_data: Eth1Data, spec: &ChainSpec) -> Self {
-        BeaconState {
+        BeaconState::Base(BeaconStateBase {
```
I noticed that `BeaconState` doesn't seem to have a `new` constructor for Altair whilst `BeaconBlock` does:

https://github.com/sigp/lighthouse/blob/1a9c89944dd06b7976a5e10bc4c8697877cea713/consensus/types/src/beacon_block.rs#L60-L66

I'm not far enough in to understand the ramifications, but I'm just making a note of it for now.

michaelsproul replied:
The reason it's currently like this is that we were creating genesis states for Altair by first making a phase0 state and then upgrading it immediately using `upgrade_to_altair`. This was a bit unwieldy, and the spec has since been updated to clarify genesis initialisation for Altair-only nets: https://github.com/ethereum/eth2.0-specs/pull/2323

When we implement that update from the spec we may well need a `BeaconState::new` that works for Altair

michaelsproul committed:
Merge remote-tracking branch 'origin/unstable' into altair

paulhauner committed:
Add tests for decoding state/block with bad slot (#2341)

paulhauner commented on beacon_node/eth2_libp2p/src/rpc/protocol.rs:
```diff
@@ -28,16 +28,16 @@ use types::{BeaconBlock, EthSpec, Hash256, MainnetEthSpec, Signature, SignedBeac
-        signature: Signature::empty(),
-    }
+    pub static ref SIGNED_BEACON_BLOCK_MIN: usize = SignedBeaconBlock::<MainnetEthSpec>::from_block(
+        BeaconBlock::<MainnetEthSpec>::empty(&MainnetEthSpec::default_spec()),
```
```suggestion
        BeaconBlock::empty(&MainnetEthSpec::default_spec()),
```

I checked this locally and I'm confident it'll compile.

paulhauner commented on testing/ef_tests/src/handler.rs:
```diff
@@ -169,8 +256,16 @@ impl<E: EthSpec + TypeName> Handler for SanityBlocksHandler<E> {
+
+    fn is_enabled_for_fork(&self, _fork_name: ForkName) -> bool {
+        // FIXME(altair): v1.1.0-alpha.3 doesn't mark the historical blocks test as
+        // requiring real crypto, so only run these tests with real crypto for now.
```
```suggestion
        // requiring real crypto, so only run these tests with fake crypto for now.
```

michaelsproul replied:
The comment is right, these tests can only be run with real crypto because they do real signature aggregation to compute sync committees. The `cfg` is `not(feature = "fake-crypto")`, i.e. real.

https://github.com/sigp/lighthouse/blob/967d680a9801bcb83676f8cdb6825b23580148cc/testing/ef_tests/src/handler.rs#L261-L263

michaelsproul replied:
Ohh I just saw the ambiguous interpretation along the lines of

"we don't _have_ to use real crypto, so use fake"

when what was meant was:

"we need to use real crypto but the tests don't state that, so temporarily hardcode it"

michaelsproul replied:
I reckon we leave it as-is for now and see how things change in alpha.4 (I'll open an issue on the spec if it's not fixed)

paulhauner commented on consensus/types/src/sync_committee.rs:
```diff
@@ -0,0 +1,32 @@
+}
+
+impl<T: EthSpec> SyncCommittee<T> {
+    /// Create a temporary sync committee that should *never* be used.
```
```suggestion
    /// Create a temporary sync committee that should *never* be included in a legitimate consensus object.
```

paulhauner commented on consensus/types/src/chain_spec.rs:
```diff
@@ -1,14 +1,22 @@
+//! This file contains several different representations of the beacon chain configuration
+//! parameters.
+//!
+//! Arguably the most import of these is `ChainSpec`, which is used throughout Lighthouse as the
```
```suggestion
//! Arguably the most important of these is `ChainSpec`, which is used throughout Lighthouse as the
```

paulhauner commented on consensus/types/src/consts.rs:
```diff
@@ -0,0 +1,21 @@
+    pub const SYNC_REWARD_WEIGHT: u64 = 8;
+    pub const PROPOSER_WEIGHT: u64 = 8;
+    pub const WEIGHT_DENOMINATOR: u64 = 64;
+    pub const INACTIVITY_SCORE_BIAS: u64 = 4;
```
I notice that `INACTIVITY_SCORE_BIAS` and `INACTIVITY_PENALTY_QUOTIENT_ALTAIR` are present here and in `ChainSpec` but I'm yet to figure out why.

michaelsproul replied:
ooh I don't think there's any good reason for that, we can probably consolidate

paulhauner committed:
Use graffiti len more consistently (#2348)

paulhauner committed:
Replace a test deleted during Altair update (#2351)

michaelsproul committed:
Apply comment fixes from Paul

Co-authored-by: Paul Hauner <paul@paulhauner.com>

paulhauner committed:
Avoid slicing in PartialBeaconState (#2361)

paulhauner committed:
[Altair] Make anonymous var-length SSZ decoding explicit (#2362)

* Add explicit function for anon var length items

* Remove unused import

* Update comment

michaelsproul committed:
[Altair] Rewards tests (#2349)

* Rewards tests (base-only so far)

* Altair rewards tests, clean-ups

* Bump Rust version in Dockerfile

paulhauner committed:
[Altair] Remove duplicate vars, panic on `u64::pow` overflow (#2350)

paulhauner commented on beacon_node/beacon_chain/src/test_utils.rs:
```diff
@@ -415,14 +444,68 @@ where
 
+    /// Useful for the `per_block_processing` tests. Creates a block, and returns the state after
+    /// caches are built but before the generated block is processed.
+    pub fn make_block_return_original_state(
```
We use `pre_state` to refer to this state in other parts of the code (it also seems like a common term outside of LH too) so perhaps `make_block_return_pre_state` would be more consistent? Just a nit really.

michaelsproul replied:
Fixed in https://github.com/sigp/lighthouse/pull/2279/commits/6415962290a54c00289cb526aefc5a304e9e94e4

paulhauner commented on consensus/ssz_derive/src/lib.rs:
```diff
@@ -152,6 +153,72 @@ pub fn ssz_encode_derive(input: TokenStream) -> TokenStream {
+///
+/// Only supports:
+/// - Enums with a single field per variant, where
+///     - All fields are variably sized from an SSZ-perspective (not fixed size).
```
Out of interest, is there a reason (apart from simplicity) why they must all be variable-sized rather than consistently variable/fixed-size?

paulhauner replied:
Hmm perhaps they don't even need to be consistent :thinking:

paulhauner replied:
> Hmm perhaps they don't even need to be consistent thinking

Oh no, `Encode::is_ssz_fixed_len()` won't allow for this.

michaelsproul replied:
> Out of interest, is there a reason (apart from simplicity) why they must all be variable-sized rather than consistently variable/fixed-size?

Just simplicity. I figured all of our enums would be for variable-sized `superstructs`. I don't think it would be worth adding support for enums with equal fixed-size variants before we have some use for them (and I doubt we would).

paulhauner committed:
[Altair] Applying Base block to Altair chain (#2353)

* Add hack to allow harness with altair spec

* Add test for applying base block to altair chain

* Add fork_name methods

* Add fork checks and tests

* Add tests for applying altair to base

* Allow providing a custom spec to test harness

* Fix compile error in network

* Add tests for slot, epoch processing

* Fix clippy lints

* Fix release test

* Fix test compile error

* Fix test with inconsistent spec

* Use fork_name_at_slot in genesis

paulhauner committed:
[Altair] Add check for missed EF tests (#2346)

* Rewards tests (base-only so far)

* Altair rewards tests, clean-ups

* Bump Rust version in Dockerfile

* Add file access logging

* Tidy

* Fix clippy lint

* Fix typo

* Tidy, remove commented-out test exceptions

* Improve naming in python script

Co-authored-by: Michael Sproul <michael@sigmaprime.io>

paulhauner committed:
Merge branch 'unstable' of github.com:sigp/lighthouse into altair-merge-unstable

michaelsproul committed:
Merge pull request #2365 from paulhauner/altair-merge-unstable

[Altair] Merge unstable

michaelsproul committed:
Restore lost CI YAML

paulhauner commented on beacon_node/operation_pool/src/lib.rs:
```diff
@@ -246,13 +250,13 @@ impl<T: EthSpec> OperationPool<T> {
         &self,
         state: &BeaconState<T>,
-        spec: &ChainSpec,
+        _spec: &ChainSpec,
```
This `_spec` is unused. Not a big deal, just pointing it out. Perhaps it's a future boi.

michaelsproul replied:
Fixed in https://github.com/sigp/lighthouse/pull/2279/commits/6415962290a54c00289cb526aefc5a304e9e94e4

paulhauner commented on consensus/state_processing/src/common/get_attestation_participation.rs:
```diff
@@ -0,0 +1,46 @@
+    let mut participation_flag_indices = SmallVec::new();
+    if is_matching_head
+        && is_matching_target
+        && state.slot() <= data.slot.safe_add(spec.min_attestation_inclusion_delay)?
```
```suggestion
        && state.slot() == data.slot.safe_add(spec.min_attestation_inclusion_delay)?
```

Fixes a slight difference from the [spec](https://github.com/ethereum/eth2.0-specs/blob/v1.1.0-alpha.3/specs/altair/beacon-chain.md#modified-process_attestation).

I don't think my suggested change would *actually* make a change at runtime, since attestations from prior to the inclusion distance will be filtered out before hitting this function. That goes for both `per_block_processing` and `operation_pool.`

michaelsproul replied:
Resolved in https://github.com/sigp/lighthouse/pull/2279/commits/6415962290a54c00289cb526aefc5a304e9e94e4

paulhauner commented on consensus/state_processing/src/common/get_attestation_participation.rs:
```diff
@@ -0,0 +1,46 @@
+    // Matching roots.
+    // Source match is checked by `verify_attestation_for_block_inclusion`.
+    let is_matching_head = data.beacon_block_root == *state.get_block_root(data.slot)?;
+    let is_matching_source = true;
```
Why don't we check the source here? I haven't been able to figure it out yet, perhaps it's something do with the op pool?

If it's just for efficiency, perhaps the gains aren't worth the risk of something "falling through the cracks"?

michaelsproul replied:
Yeah, you're right that it doesn't hurt to have that check in there. This function existed as a sort of weird intermediate that I had to yank out of attestation processing, but it's now been refactored in the spec into a real entity: https://github.com/ethereum/eth2.0-specs/blob/v1.1.0-alpha.6/specs/altair/beacon-chain.md#get_attestation_participation_flag_indices

I'll fix it on this branch, and then carry that over when we do the update

michaelsproul replied:
Resolved in https://github.com/sigp/lighthouse/pull/2279/commits/6415962290a54c00289cb526aefc5a304e9e94e4

michaelsproul committed:
Address review comments

michaelsproul committed:
Update comment

paulhauner committed:
Return an error if total bal is 0 (#2366)

paulhauner commented on consensus/state_processing/src/per_epoch_processing/weigh_justification_and_finalization.rs:
```diff
@@ -0,0 +1,67 @@
+        *state.finalized_checkpoint_mut() = old_previous_justified_checkpoint;
+    }
+    // The 2nd/3rd most recent epochs are both justified, the 2nd using the 3rd as source.
+    else if (1..3).all(|i| bits.get(i).unwrap_or(false))
```
```suggestion
    if (1..3).all(|i| bits.get(i).unwrap_or(false))
```

This is a non-substantive change, but it makes our code look more like the spec.

paulhauner replied:
Addressed in https://github.com/sigp/lighthouse/pull/2367

paulhauner commented on consensus/state_processing/src/per_epoch_processing/weigh_justification_and_finalization.rs:
```diff
@@ -0,0 +1,67 @@
+        *state.finalized_checkpoint_mut() = old_current_justified_checkpoint;
+    }
+    // The 1st/2nd most recent epochs are both justified, the 1st using the 2nd as source.
+    else if (0..2).all(|i| bits.get(i).unwrap_or(false))
```
```suggestion
    if (0..2).all(|i| bits.get(i).unwrap_or(false))
```

As per previous comment regarding looking more like the spec.

paulhauner replied:
Addressed in https://github.com/sigp/lighthouse/pull/2367

paulhauner committed:
Minor changes to justification code (#2367)

paulhauner commented on consensus/state_processing/src/per_epoch_processing/altair/participation_flag_updates.rs:
```diff
@@ -0,0 +1,21 @@
+        std::mem::take(state.current_epoch_participation_mut()?);
+    *state.current_epoch_participation_mut()? = VariableList::new(vec![
+        ParticipationFlags::default(
+        );
```
Hmm, looks like a bug in `rustfmt`. No change required, just pointing it out. I tried to put the `);` back onto L16 but `rustfmt` put it back :shrug:

michaelsproul replied:
How weird!

michaelsproul commented on consensus/types/src/participation_flags.rs:
```diff
@@ -0,0 +1,87 @@
+
+impl ParticipationFlags {
+    pub fn add_flag(&mut self, flag_index: u32) -> Result<(), ArithError> {
+        if flag_index > NUM_FLAG_INDICES as u32 {
```
Just noticed this should be a `>=`. Fixing it in the alpha.5/6 branch, but can back port it as necessary.

michaelsproul commented on consensus/types/src/participation_flags.rs:
```diff
@@ -0,0 +1,87 @@
+    }
+
+    pub fn has_flag(&self, flag_index: u32) -> Result<bool, ArithError> {
+        if flag_index > NUM_FLAG_INDICES as u32 {
```
Off-by-one, as per above

paulhauner reviewed (approved):
After my longest and most detailed review *ever*, I'm very excited to add this little green tick. I spent quite a lot of time reviewing this code and I don't believe I found *any* consensus bugs. I did find interesting nuances and improvements so I'm satisfied that my review process was thorough. If there *are* bugs remaining, my approval means that I share in the responsibility equally. (Of course, no *legal* responsibility is assumed by *anyone* and we remain fully protected by our Apache 2.0 license).

Great work @michaelsproul and @realbigsean, this is an awesome effort. Not only did this include the Altair logic, it significantly improved our existing implementation through efforts like removing array-slicing. I got RSI by just *looking* at all the changes from direct field access to getter/setters. Very, very well done! :champagne: :clinking_glasses: 

As discussed with @michaelsproul, I think we should hold off on merging this until we cut another minor release. I also have [a PR](https://github.com/sigp/lighthouse/pull/2369) that's targeting this branch.

I suspect that we'll merge a few more things into this PR over time. We can do re-reviews as this happens.

paulhauner committed:
[Altair] Optimization for `get_unslashed_participating_indices` (#2369)

* Lift calculation of unslashed indices

* Use HashSet for unslashed participating indices

* Make Operation Pool Compile Again (MOPCA)

michaelsproul committed:
Update consensus code to v1.1.0-alpha.6

michaelsproul committed:
Ripple spec updates

michaelsproul committed:
Implement preset <> config distinction

michaelsproul committed:
Return ConfigAndPreset from VC API

michaelsproul committed:
Don't panic if there aren't any aggregators

Rustfmt also returned with a vengeance to rearrange this block.

michaelsproul committed:
Check state root in transition tests

michaelsproul committed:
Fix clippy

michaelsproul committed:
Add compatibility shim for /config/spec

michaelsproul committed:
Fix Pyrmont & Prater configs. Delete Toledo.

michaelsproul commented:
Will fix merge conflicts here shortly before merging (in order to reduce churn in all the downstream PRs)

michaelsproul committed:
Minor cleanups

michaelsproul committed:
Merge pull request #2373 from michaelsproul/altair-latest-spec

[Altair] Update to latest spec v1.1.0-alpha.6

michaelsproul committed:
Update to spec v1.1.0-alpha.7

michaelsproul committed:
Update mainnet preset YAML

michaelsproul committed:
Merge pull request #2405 from sigp/altair-alpha.7

Update to spec v1.1.0-alpha.7

michaelsproul committed:
Merge remote-tracking branch 'origin/unstable' into altair

michaelsproul committed:
Fix beacon_chain test

realbigsean committed:
Updates related to merge with `unstable`

realbigsean committed:
Couple more `unstable` merge fixes

michaelsproul committed:
Merge remote-tracking branch 'origin/unstable' into altair

michaelsproul committed:
Merge remote-tracking branch 'origin/unstable' into altair

michaelsproul committed:
Merge remote-tracking branch 'origin/unstable' into altair

michaelsproul committed:
Merge remote-tracking branch 'origin/unstable' into altair

paulhauner reviewed (approved):
Merge it!

michaelsproul commented:
Can't quite believe it... let's try this

bors r+

:tada:
